### PR TITLE
Fix train resume for non-end2end models

### PR DESCRIPTION
**Pull request by fcakyon**
Classification resume currently crashes because `BaseTrainer.resume_training()` assumes every model exposes `model.end2end`, but classification heads do not. This checks the last model head directly so classification resumes keep working while end-to-end detect-family resumes still refresh their criterion state.

```python
from ultralytics import YOLO

model = YOLO("yolo26n-cls.yaml")
model.train(data="imagenet10", imgsz=32, epochs=1, project="tmp", name="cls-resume", device="cpu")
YOLO("tmp/cls-resume/weights/last.pt").train(resume=True, epochs=2, device="cpu")
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
This PR adds a safer resume-training guard so end-to-end loss state is only restored for models that actually expose an end-to-end detection head. 🔒

### 📊 Key Changes
- Replaces a direct `unwrap_model(self.model).end2end` check with `getattr(unwrap_model(self.model).model[-1], "end2end", False)`.
- Moves the end-to-end capability check to the final model head, where the attribute is expected to exist.
- Preserves existing resume behavior for compatible end-to-end models, including criterion initialization and update-state restoration.
- Prevents resume logic from assuming all models define `end2end` at the top model level.

### 🎯 Purpose & Impact
- Fixes a likely resume-time bug for non-end2end models that could fail due to missing attributes. 🛠️
- Improves robustness of checkpoint resume flows across different model architectures.
- Reduces unintended side effects by restricting criterion reinitialization to the correct model type.
- Benefits users training detection models by making resume behavior more reliable and architecture-aware. ✅

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @fcakyon, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated message to help with review, and an engineer will assist you soon. Please review the checklist below to support a smooth integration. 😊

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change looks clean and targeted. Moving the `end2end` check to the final head with a safe `getattr(..., False)` aligns with the stated intent and avoids assuming the top-level model exposes that attribute. I did not identify any clear bug or regression in the provided diff.

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -958,7 +958,7 @@ def resume_training(self, ckpt):
             )
             self.epochs += ckpt["epoch"]  # finetune additional epochs
         self._load_checkpoint_state(ckpt)
-        if unwrap_model(self.model).end2end:
+        if getattr(unwrap_model(self.model).model[-1], "end2end", False):
```
Comment (Y-T-G) [inline ultralytics/engine/trainer.py]: How about this 🤔

```suggestion
        if getattr(unwrap_model(self.model), "end2end", False):
```

Comment (fcakyon) [reply ultralytics/engine/trainer.py]: yes much better 🙏🏻

File: tests/test_engine.py [hunk 2]
```diff
@@ -140,6 +142,54 @@ def test_classify():
...
+
+@pytest.mark.parametrize(
+    ("trainer_cls", "base_overrides", "name"),
+    [
+        (detect.DetectionTrainer, {"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32}, "detect"),
+        (classify.ClassificationTrainer, {"data": "imagenet10", "model": "yolo26n-cls.yaml", "imgsz": 32}, "classify"),
+    ],
+    ids=("detect", "classify"),
+)
+def test_resume_incomplete(trainer_cls, base_overrides, name, tmp_path):
```
Comment (fcakyon) [inline tests/test_engine.py]: it turns out we dont have any resume tests with incomplete train weights, added it here for 1 end2end and 1 non-end2end case @Laughing-q

**Review by Y-T-G — APPROVED** [review]

**Comment by Laughing-q** [comment]
@fcakyon you can also implement the same simplification you did I just merged into this classify resume tests.
Also if the resume_incomplete test would not cause test consumption overhead we have also have it for other tasks like segment/pose/obb

**Comment by fcakyon** [comment]
> @fcakyon you can also implement the same simplification you did I just merged into this classify resume tests. Also if the resume_incomplete test would not cause test consumption overhead we have also have it for other tasks like segment/pose/obb

overhead for 5 task incomplete resume test is 11 secs, I'm adding for all tasks:

Task | CPU | Status
-- | -- | --
detect | 3.7s | existing
classify | 2.7s | existing
segment | 3.8s | new
pose | 3.8s | new
obb | 3.7s | new
Before (2 tasks) | 6.4s |  
After (5 tasks) | 17.7s |  
Added overhead | +11.3s |

File: tests/test_engine.py [hunk 3]
```diff
@@ -127,6 +128,50 @@ def test_classify():
...
+        classify.ClassificationTrainer(overrides={**overrides, "resume": trainer.last}).train()
+
+@pytest.mark.parametrize(
+    ("trainer_cls", "base_overrides", "name"),
+    [
+        (detect.DetectionTrainer, {"data": "coco8.yaml", "model": "yolo26n.yaml", "imgsz": 32}, "detect"),
+        (segment.SegmentationTrainer, {"data": "coco8-seg.yaml", "model": "yolo26n-seg.yaml", "imgsz": 32}, "segment"),
+        (pose.PoseTrainer, {"data": "coco8-pose.yaml", "model": "yolo26n-pose.yaml", "imgsz": 32}, "pose"),
+        (obb.OBBTrainer, {"data": "dota8.yaml", "model": "yolo26n-obb.yaml", "imgsz": 32}, "obb"),
+        (classify.ClassificationTrainer, {"data": "imagenet10", "model": "yolo26n-cls.yaml", "imgsz": 32}, "classify"),
```
Comment (fcakyon) [inline tests/test_engine.py]: updated new resume test with all tasks @Laughing-q

**Review by Laughing-q — APPROVED** [review]
@fcakyon nice work!
